hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -442,6 +450,81 @@ def runTest(self):
         StateMachineTestCase.__qualname__ = cls.__qualname__ + ".TestCase"
         return StateMachineTestCase
 
+    def _execute_fn(self, rule, stateful_run_times, **data):
+        result = get_sync_result_with_time(
+            rule.function, stateful_run_times, self, **data
+        )
+        return result
+
+
+class AsyncIORuleBasedStateMachine(RuleBasedStateMachine):
+    def __init__(self):
+        super().__init__()
+        global asyncio
+        import asyncio
+
+    def _execute_fn(self, rule, stateful_run_times, **data):
+        if inspect.iscoroutinefunction(rule.function):
+            # Make sure to wrap it in get_async_result_with_time once debugged
+            self.tg.create_task(rule.function(self, **data))
+            return
```

Comment:
This is fundamentally not going to work - at minimum, it's concurrency between and not just within rules.

I think it's probably time to step back and write up a detailed design, rather than jumping in on implementation - in the linked issue I'm not even sure what a working async stateful test _should_ look like, or _could_ look like, and so I'm concerned that we'll end up thrashing a lot because it's also very difficult to review such intricate code.

Reply:
I actually designed it with the concurrency between rules in mind (specifically the "keeping a single context manager open for the duration"), though I didn't really update my description for this. 

I agree though that we should concretely figure out the design given the complexity. I'll work on that!